src/riak_kv_wm_preflist.erl

```diff
@@ -0,0 +1,226 @@
+%%      Properties allows, GET.
+allowed_methods(RD, Ctx) ->
+    {['GET'], RD, Ctx}.
+
+-spec content_types_provided(#wm_reqdata{}, context()) ->
+    {[{ContentType::string(), Producer::atom()}], #wm_reqdata{}, context()}.
+%% @doc List the content types available for representing this resource.
+%%      "application/json" is the content-type for listing keys.
+content_types_provided(RD, Ctx) ->
+    %% bucket-level: JSON description only
+    {[{"application/json", produce_preflist_body}], RD, Ctx}.
+
+-spec encodings_provided(#wm_reqdata{}, context()) ->
+    {[{Encoding::string(), Producer::function()}], #wm_reqdata{}, context()}.
+%% @doc List the encodings available for representing this resource.
+%%      "identity" and "gzip" are available for a preflist request.
+encodings_provided(RD, Ctx) ->
+    %% identity and gzip for top-level and bucket-level requests
+    {riak_kv_wm_utils:default_encodings(), RD, Ctx}.
+
+resource_exists(RD, #ctx{bucket_type=BType}=Ctx) ->
+    {riak_kv_wm_utils:bucket_type_exists(BType), RD, Ctx}.
+
+-spec malformed_request(#wm_reqdata{}, context()) ->
+    {boolean(), #wm_reqdata{}, context()}.
+malformed_request(RD, Ctx) ->
+    {false, RD, Ctx}.
+
+-spec produce_preflist_body(#wm_reqdata{}, context()) ->
+                                   {mochijson2:json_object()|{halt, 404},
+                                    #wm_reqdata{},
+                                    context()}.
+%% @doc Produce bucket/key annotated preflist as JSON.
+produce_preflist_body(RD, #ctx{bucket=Bucket0,
+                               bucket_type=BType,
+                               key=Key}=Ctx) ->
+    Bucket = riak_kv_wm_utils:maybe_bucket_type(BType, Bucket0),
+    Preflist = riak_core_apl:get_apl_ann_with_pnum({Bucket, Key}),
+    case Preflist of
+        [] ->
+            {{halt, 404},
+             wrq:set_resp_header("Content-Type", "text/plain",
+                 wrq:append_to_response_body(
+                     io_lib:format(
+                       "No preflist for bucket '~s' and key '~s'~n",
+                       [Bucket, Key]),
+                   RD)),
+             Ctx};
+        P ->
+            %% Encode
+            Json = mochijson2:encode({struct,
+                                      [{<<"preflist">>,
+                                        lists:flatten(jsonify_results(P))}]}),
+            {Json, RD, Ctx}
+    end.
+
+%% Private
+
+-spec jsonify_results(riak_core_apl:preflist_with_pnum_ann()) -> list().
+%% @doc Jsonify active preflist to json.
+jsonify_results(Preflist) ->
+    [jsonify_results({PartitionNumber, Node}, T) ||
+        {{PartitionNumber, Node}, T} <- Preflist].
+
+-spec jsonify_results({non_neg_integer(), node()}, primary|fallback) ->
+                            [{struct, list()}].
+
+jsonify_results({PartitionNumber, Node}, primary) ->
+    [{struct,
+      [{<<"partition">>, PartitionNumber},
+       {<<"node">>, atom_to_binary(Node, utf8)},
+       {<<"primary">>, true}]}]
```

Comment:
You could collapse this function to a single clause if you replace `primary` in the function head with `Ann` and use `Ann == primary` in the field. I would also call this function `jsonify_result` or `jsonify_preflist_entry` or something to distinguish it from the 1-arity list-taking function above.